Implement - Add pre-commit hook to automatically run frontend formatting (npm run format) #2259

sagarkhandagre998 opened the pull request:
Implementation- 
Add pre-commit hook to automatically format frontend code before each commit.

---

**Changes**
- Installed **Husky (v9.1.7)** as devDependency
- Added `prepare` script to auto-initialize Husky on `npm install`
- Created `.husky/pre-commit` hook: runs `cd frontend && npm run lint-staged`
- Configured lint-staged** following  pattern:
   - ESLint + Prettier for JS/JSX/TS/TSX files
   - Prettier only for CSS/SCSS/JSON/MD files
   - Only processes staged files
- Updated `CONTRIBUTING.md` with setup instructions

---

### ✅ **How It Works**
1. Developer commits code: `git commit -m "message"`
2. Pre-commit hook automatically runs `npm run lint-staged`
3. Formatted files that actually changed/created are staged automatically
4. Commit proceeds

---

### 🔧 **Setup (One-time)**
```bash
cd frontend
npm install
```

Hook is now active and will run automatically on every commit.

---

### ✅ **Tested**
- Created test file with unformatted code
- Committed successfully
- File was automatically formatted before commit

sagarkhandagre998 committed:
feat: add pre-commit hook for automatic frontend formatting

- Install Husky (v9.1.7) as devDependency
- Add prepare script to initialize Husky on npm install
- Create .husky/pre-commit hook that runs npm run format and stages changes
- Update CONTRIBUTING.md with setup instructions

Resolves issue by implementing automatic code formatting before each commit
as specified in the issue description.

sagarkhandagre998 committed:
Implements automatic frontend code formating whith husky

sagarkhandagre998 committed:
remove the testing code

sagarkhandagre998 committed:
apply mvn code formating

sagarkhandagre998 committed:
fix: skip Husky initialization in production builds

Add fallback to prepare script to prevent Docker build failures when
devDependencies are omitted in production.

Agaba-derrick commented on frontend/package.json:
```diff
@@ -102,6 +103,7 @@
     "eslint-plugin-prettier": "^5.0.0",
```
does this make sense in terms of formating only staged files ?
https://github.com/openmrs/openmrs-esm-patient-chart/blob/main/package.json#L83C2-L86C5

sagarkhandagre998 replied:
Yeah definitely, It will not format all files. Format only the files that developer has changed or implemented some new files,  so after commiting (all the chagned or new files) will get automatically format so we don't need to format all. Thats why i previously thought to implement this.

Agaba-derrick replied:
are you able to make  your changes look like those referenced  from that block i have shared ?

sagarkhandagre998 replied:
Sure I'll do that, Need Some time to analyze. I'll commit changes soon.

sagarkhandagre998 replied:
> are you able to make your changes look like those referenced from that block i have shared ?

But they only {ts,tsx} files in their particular package but in our frontend we need to include {js ,jsx ,ts ,tsx} and {css,scss,json,md} also. I'll make sure to include this as well.

sagarkhandagre998 committed:
refactor: use lint-staged to format only staged files with ESLint + Prettier

Replace 'npm run format' (formats all files) with lint-staged configuration
following maintainer's recommended pattern:

- Run ESLint with auto-fix and zero warnings on JS/JSX/TS/TSX files
- Run Prettier on all staged files (JS/JSX/TS/TSX/CSS/SCSS/JSON/MD)
- Only process staged files (faster, safer)
- Add .eslintcache to .gitignore

Benefits:
- Faster commits (only formats changed files, not all 378 files)
- Enforces code quality (ESLint with --max-warnings 0)
- Prevents staging unintended files
- Follows project's ESLint + Prettier pattern

sagarkhandagre998 committed:
Update the pre-commit hook to use the locally installed version and adding the lint-staged script to package.json

sagarkhandagre998 committed:
warning checks working fine, not allowing to commit if changes have a warning - good behavior

sagarkhandagre998 committed:
removed file from gitignore

sagarkhandagre998 committed:
need to check where .eslintcache file is created ?

sagarkhandagre998 committed:
fixing gitignore file

sagarkhandagre998 committed:
Merge branch 'develop' into issue-2259

Agaba-derrick commented:
@sagarkhandagre998  Thanks, we are clossing this PR as this issue was resolved in some earlier work by @pmanko
